### PR TITLE
Remove most trait implementation features from `esp-hal`

### LINKED ISSUE #2007: Remove or enable the `async` feature
By default. We previously had this as a feature because of the way we were doing interrupts meant that there were linker conflicts. Now that we do runtime binding of interrupts, we could enable this by default, or ditch the feature entirely. The only argument I can think of for keeping it is for compile time reasons, but I can't see it making a big dent overall.

### PR DESCRIPTION
**Pull request by jessebraham**
I accidentally nuked my branch and had to close #2047, but I believe this should now be in a good state again. See the previous PR for conversation, I believe the review comments have been addressed.

Adding `skip-changelog` again to silence the other packages, but `esp-hal` will have its `CHANGELOG.md` updated still.

Closes #2007

### CODE REVIEW

**Comment by bugadani** [comment]
We might want to mention this in the shiny new migration guide ;)

File: esp-hal/MIGRATING-0.20.md [hunk 1]
```diff
@@ -13,7 +24,7 @@ Instead of manually grabbing peripherals and setting up clocks, you should now c
      prelude::*,
 -    system::SystemControl,
  };
- 
+
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.20.md:27]: Technically this is now invalid diff sytanx 🙃

File: esp-hal/MIGRATING-0.20.md [hunk 2]
```diff
@@ -1,8 +1,19 @@
-Migration Guide from 0.20.x to vNext
-====================================
+# Migration Guide from 0.20.x to vNext
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.20.md:1]: Aww

**Review by bugadani — APPROVED** [review]

File: esp-hal/MIGRATING-0.20.md (hunk 2 above)
Comment (jessebraham) [reply esp-hal/MIGRATING-0.20.md:1]: Editor has format on save enabled, all the other markdown files are like this. Can revert if needed but I would rather not.

File: esp-hal/MIGRATING-0.20.md (hunk 1 above)
Comment (jessebraham) [reply esp-hal/MIGRATING-0.20.md:27]: How so?

File: esp-hal/MIGRATING-0.20.md (hunk 2 above)
Comment (bugadani) [reply esp-hal/MIGRATING-0.20.md:1]: Nah I'll not fight The Formatter, I'm just sad :)

File: esp-hal/MIGRATING-0.20.md (hunk 1 above)
Comment (bugadani) [reply esp-hal/MIGRATING-0.20.md:27]: AFAIK the first character in a diff denotes what's going on, where a space means "unchanged line". It's just a markdown, so there's no difference here, but git might complain (if we cared about that here).

**Review by bjoernQ — APPROVED** [review]
Thanks

**Pull request merged**